src/solvers/refinement/string_refinement.cpp

```diff
@@ -0,0 +1,769 @@
+      debug() << "universaly quantified : " << pretty_short(generator.axioms[i]) << eom;
+      universal_axioms.push_back(generator.axioms[i]);
+    }
+    else {
+      assert(generator.axioms[i].is_not_contains());
+      generator.axioms[i].witness = string_exprt::fresh_symbol
+	("not_contains_witness",
+	 array_typet(refined_string_typet::index_type(),
+		     infinity_exprt(refined_string_typet::index_type())));
+      not_contains_axioms.push_back(generator.axioms[i]);
+    }
+
+  //string_axioms.clear(); should not be necessary
+
+  initial_index_set(universal_axioms);
+  debug() << "string_refinementt::dec_solve: warning update_index_set has to be checked" << eom;
+  update_index_set(cur);
+  cur.clear();
+  add_instantiations();
+
+  while(initial_loop_bound-- > 0)
+    {
+
+      print_time("string_refinementt::dec_solve");
+      decision_proceduret::resultt res = SUB::dec_solve();
+
+      switch(res)
+	{
+	case D_SATISFIABLE:
+	  if(!check_axioms()) {
+	    debug() << "check_SAT: got SAT but the model is not correct" << eom;
+	  }
+	  else {
+	    debug() << "check_SAT: the model is correct" << eom;
+	    return D_SATISFIABLE;
+	  }
+
+	  debug() <<  "refining.." << eom;
+	  current_index_set.clear();
+	  update_index_set(cur);
+	  cur.clear();
+	  add_instantiations();
+
+	  if(variable_with_multiple_occurence_in_index)
+	    {
+	      debug() << "WARNING: some variable appears multiple times" << eom;
+	    }
+
+	  if(current_index_set.empty()){
+	    debug() << "current index set is empty" << eom;
+	    return D_SATISFIABLE;
+	  }
+
+	  display_index_set();
+	  debug()<< "instantiating NOT_CONTAINS constraints" << eom;
+	  for(unsigned i=0; i<not_contains_axioms.size(); i++) {
+	    debug()<< "constraint " << i << eom;
+	    std::vector<exprt> lemmas;
+	    instantiate_not_contains(not_contains_axioms[i],lemmas);
+	    for(unsigned j=0; j<lemmas.size(); j++) {
+	      add_lemma(lemmas[j]);
+	    }
+	  }
+	  break;
+	default:
+	  return res;
+	}
+
+    }
+  debug () << "string_refinementt::dec_solve reached the maximum number of steps allowed";
+  return D_ERROR;
+}
+
+
+bvt string_refinementt::convert_bool_bv(const exprt &boole, const exprt &orig)
+{
+  bvt ret;
+  ret.push_back(convert(boole));
+  size_t width = boolbv_width(orig.type());
+  for (size_t i = 1; i < width; ++i) {
+    ret.push_back(const_literal(false));
+  }
+  return ret;
+}
+
+void string_refinementt::add_lemma(const exprt &lemma, bool add_to_index_set)
+{
+  if (!seen_instances.insert(lemma).second)
+    {
+      debug() << "string_refinementt::add_lemma : already seen" << eom;
+      return;
+    }
+
+  if(lemma == true_exprt())
+    {
+      debug() << "string_refinementt::add_lemma : tautology" << eom;
+      return;
+    }
+
+  debug() << "adding lemma " << pretty_short(lemma) << eom;
+
+  prop.l_set_to_true(convert(lemma));
+  if(add_to_index_set)
+    cur.push_back(lemma);
+}
+
+unsigned integer_of_expr(const constant_exprt & expr)
```

Comment:
Done in https://github.com/diffblue/cbmc/pull/341/commits/2ab3e49e3a552a59eab3c1b54e104f3d49161f93